internal/hostname/providers.go

```diff
@@ -0,0 +1,243 @@
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/hostname/validate"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+// For testing purposes
+var (
+	fargatePf = fargate
+)
+
+var (
+	cachedHostname  string
+	cachedAt        time.Time
+	cachedProvider  string
+	cacheExpiration = 5 * time.Minute
+	m               sync.RWMutex
+	isRefreshing    atomic.Value
+)
+
+const fargateName = "fargate"
+
+func init() {
+	isRefreshing.Store(false)
+}
+
+// getCached returns the cached hostname, cached provider and a bool indicating if the hostname has expired
+func getCached(now time.Time) (string, string, bool) {
+	m.RLock()
+	defer m.RUnlock()
+	if now.Sub(cachedAt) > cacheExpiration {
+		return cachedHostname, cachedProvider, true
+	}
+	return cachedHostname, cachedProvider, false
+}
+
+// setCached caches the newHostname
+func setCached(now time.Time, newHostname string, newProvider string) {
+	m.Lock()
+	defer m.Unlock()
+	cachedHostname = newHostname
+	cachedAt = now
+	cachedProvider = newProvider
+}
+
+type provider struct {
+	name string
+	// Should we stop going down the list of providers if this one is successful
+	stopIfSuccessful bool
+	pf               providerFetch
+}
+
+type providerFetch func(ctx context.Context, currentHostname string) (string, error)
+
+var providerCatalog = []provider{
+	{
+		name:             "configuration",
+		stopIfSuccessful: true,
+		pf:               fromConfig,
+	},
+	{
+		name:             fargateName,
+		stopIfSuccessful: true,
+		pf:               fromFargate,
+	},
+	{
+		name:             "gce",
+		stopIfSuccessful: true,
+		pf:               fromGce,
+	},
+	{
+		name:             "azure",
+		stopIfSuccessful: true,
+		pf:               fromAzure,
+	},
+	// The following providers are coupled. Their behavior changes depending on the result of the previous provider.
+	// Therefore, 'stopIfSuccessful' is set to false.
+	{
+		name:             "fqdn",
+		stopIfSuccessful: false,
+		pf:               fromFQDN,
+	},
+	{
+		name:             "container",
+		stopIfSuccessful: false,
+		pf:               fromContainer,
+	},
+	{
+		name:             "os",
+		stopIfSuccessful: false,
+		pf:               fromOS,
+	},
+	{
+		name:             "aws",
+		stopIfSuccessful: false,
+		pf:               fromEC2,
+	},
+}
+
+// Get returns the cached hostname for the tracer, empty if we haven't found one yet.
+// Spawning a go routine to update the hostname if it is empty or out of date
+func Get() string {
+	now := time.Now()
+	var (
+		ch      string
+		expired bool
+		pv      string
+	)
+	// if provider is fargate never refresh
+	// Otherwise, refresh on expiration or if hostname hasn't been found.
+	if ch, pv, expired = getCached(now); pv == fargateName || (!expired && ch != "") {
+		return ch
+	}
+	// Use CAS to avoid spawning more than one go-routine trying to update the cached hostname
+	ir := isRefreshing.CompareAndSwap(false, true)
+	if ir {
+		go func() {
+			updateHostname(now)
```

Comment:
`updateHostname` can get stuck for a relatively long time (its timeout is 5 minutes, L143). What if the app wants to shutdown during that timeframe?
I think a cancellation signal should be propagated from `tracer.Stop()` to this function, we should be able to cancel it immediately.

Reply:
Ooooo excellent catch here

Reply:
Actually thinking about this more, do we need to propagate a cancel signal? Normally we would do that so we can force a flush of any data that might be waiting, but in this case we are fine to just drop our work as it's happening. And since the goroutine will be cleaned up by the system once the `main` goroutine exits I'm not sure it's worth adding the extra complexity here? (I might be missing or forgetting something here though)

Reply:
I'll add a todo here :) Also the fact that this go routine creation is protected by that `CAS` luckily means that even in weird (read: bad) setups where a user starts and stops a tracer several times quickly, we won't pollute with a ton of go routines here